components/cldr-json-data-provider/src/transform/plurals.rs

```diff
@@ -0,0 +1,232 @@
+    type Error = Error;
+    fn try_from(cldr_paths: &CldrPaths) -> Result<Self, Self::Error> {
+        let cardinal_rules = {
+            let path = cldr_paths
+                .cldr_core
+                .clone()?
+                .join("supplemental")
+                .join("plurals.json");
+            let data: cldr_json::Resource = serde_json::from_reader(open_reader(path)?)?;
+            data.supplemental
+                .plurals_type_cardinal
+                .map(|r| r.normalize())
+        };
+        let ordinal_rules = {
+            let path = cldr_paths
+                .cldr_core
+                .clone()?
+                .join("supplemental")
+                .join("ordinals.json");
+            let data: cldr_json::Resource = serde_json::from_reader(open_reader(path)?)?;
+            data.supplemental
+                .plurals_type_ordinal
+                .map(|r| r.normalize())
+        };
+        Ok(PluralsProvider {
+            cardinal_rules,
+            ordinal_rules,
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'d> TryFrom<&'d str> for PluralsProvider<'d> {
+    type Error = serde_json::error::Error;
+    /// Attempt to parse a JSON string.
+    fn try_from(s: &'d str) -> Result<Self, Self::Error> {
+        let data: cldr_json::Resource = serde_json::from_str(s)?;
+        Ok(PluralsProvider {
+            cardinal_rules: data
+                .supplemental
+                .plurals_type_cardinal
+                .map(|r| r.normalize()),
+            ordinal_rules: data
+                .supplemental
+                .plurals_type_ordinal
+                .map(|r| r.normalize()),
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'d> DataKeySupport for PluralsProvider<'d> {
+    fn supports_key(data_key: &DataKey) -> Result<(), DataError> {
+        if data_key.category != DataCategory::Plurals {
+            return Err((&data_key.category).into());
+        }
+        if data_key.version != 1 {
+            return Err(data_key.into());
+        }
+        Ok(())
+    }
+}
+
+impl<'d> PluralsProvider<'d> {
+    fn get_rules_for(&self, data_key: &DataKey) -> Result<&cldr_json::Rules, DataError> {
+        PluralsProvider::supports_key(data_key)?;
+        match data_key.sub_category.as_str() {
+            "cardinal" => self.cardinal_rules.as_ref(),
+            "ordinal" => self.ordinal_rules.as_ref(),
+            _ => return Err(data_key.into()),
+        }
+        .ok_or_else(|| data_key.into())
+    }
+}
+
+impl<'d> DataProvider<'d> for PluralsProvider<'d> {
+    fn load(&self, req: &DataRequest) -> Result<DataResponse<'d>, DataError> {
+        let cldr_rules = self.get_rules_for(&req.data_key)?;
+        // TODO: Implement language fallback?
+        // TODO: Avoid the clone
+        let cldr_langid = CldrLanguage(req.data_entry.langid.clone());
+        let (_, r) = match cldr_rules.0.binary_search_by_key(&&cldr_langid, |(l, _)| l) {
```

Comment:
nit: `&&` can be confusing to look through for readers who may be unsure of `binary_search` requirement of referencing, bring the second `&` out to the tuple:
```rust
let (_, r) = match cldr_rules.0.binary_search_by_key(&cldr_langid, |&(l, _)| l) {
``` 
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by_key puts it out the tuple as well